### ISSUE TITLE
Add a way to define custom SQL DDL for Index

### ISSUE THREAD

**Issue by lukasj** [jakartaee/persistence#636]
The spec should allow user to define his own, custom DDL for Index definition to support usage of DB specific features, ie MySQL `CREATE [UNIQUE | FULLTEXT | SPATIAL] INDEX ...` or Oracle DB `CREATE [BITMAP | MULTIVALUE | VECTOR ] INDEX ...` We already have `@Column.columnDefinition`, so we could add something like `@Index.indexDefinition`

**Comment by loiclefevre** [comment]
@lukasj, there is also `CREATE SEARCH INDEX ...` for Oracle DB:

See [JSON Developer Guide](https://docs.oracle.com/en/database/oracle/oracle-database/23/adjsn/indexes-for-json-data.html#GUID-D7A604E1-F617-4C92-A9D9-00B6AB78EFD8)

**Comment by gavinking** [comment]
So at one stage a did spend a tiny amount of time thinking about this, the only issue is:

1. Just adding `options` (which was have done in 3.2) wasn't quite enough because as you say we need a way to specify some arbitrary qualifiers that go before the `index` keyword.
2. Just adding `indexDefinition` didn't seem especially useful to me because you might as well just add such things in the script specified by `jakarta.persistence.schema-generation.create-source`. Why use an `@Index` annotation at all for that?

What we _could_ do though is add a `type` member to some of these annotations, so you could write stuff like:

```java
@Table(name="TheTable", 
       indexes=@Index(name="TheIndex", 
                      type="fulltext", 
                      columnList="column1, column2",
                      options="with parser MyParser"))
```

which would result in:

```sql
create table TheTable ( ..... fulltext index TheIndex (column1, column2) with parser MyParser)
```

I think that gives you everything you need for indexes, and the `type` element also makes sense for `@Table` and I guess `@UniqueConstraint` and `@ForeignKey` as well.

**Comment by loiclefevre** [comment]
The `type` for `@Index` looks exactly like what we need. This is also needed for `@Table`; I'm not sure about the others though.

I also wanted to double-check if the `columnList` value is parsed at some point and validated according to the BNF grammar mentioned in the Javadoc?

**Comment by gavinking** [comment]
> I also wanted to double-check if the `columnList` value is parsed at some point and validated according to the BNF grammar mentioned in the Javadoc?

Good point. In our implementation we do actually parse it, yes. But if we did this, I think we could allow the `columnList` to be freeform text. It's not really clear at all what value that grammar has.

**Comment by loiclefevre** [comment]
I can see 3 types of index where validating the grammar could be problematic:
- function based indexes
- JSON Path expressions to be indexed (a special case for #1)
- multivalue index définies over JSON_Table()

**Comment by gavinking** [comment]
Fixed, see #934, #935.
